python/CuTeDSL/cutlass/cute/core.py

```diff
@@ -4111,6 +4111,168 @@ def align_offset(offset, align):
+        @cute.union
+        class aligned_union:
+            a: cute.struct.Align[cutlass.Float32, 16]
+            b: cute.struct.Align[cutlass.Int32, 8]
+
+
+        # Statically get size and alignment:
+        size = data_union.__sizeof__()
+        align = data_union.__alignof__()
+
+        # Allocate and reference elements:
+        allocator = cutlass.utils.SmemAllocator()
+        value = allocator.allocate(data_union)
+
+        # Access union members (all at the same offset):
+        value.small ...
+        value.medium ...
+        value.large ...
+
+    :param cls: The union class with annotations.
+    :return: The decorated union class.
+    """
+
+    def __init__(self, cls):
+        """
+        Initializes a new cute.union decorator instance.
+
+        :param cls: The class representing the union data type.
+        :raises TypeError: If the union is empty.
+        """
+        self._cls = cls
+        self.__name__ = f"cute.union::{cls.__name__}"
+        # Get the class annotations
+        self._annotations = getattr(cls, "__annotations__", {})
+        # Create a dictionary to store the offsets (all zeros for union)
+        self._offsets: Dict[str, int] = {}
+
+        # Calculate the maximum size and alignment
+        max_size = 0
+        max_alignment = 1
+        if len(self._annotations) == 0:
+            raise TypeError("Empty union is not supported!")
+
+        for name, object in self._annotations.items():
+            # All offsets are 0 for a union
+            self._offsets[name] = 0
+
+            # Get alignment of object
+            sub_align = 1
+            if isinstance(object, struct._AlignMeta):
+                sub_align = object.align
+                object = object.dtype
+
+            # Calculate size and alignment based on object type
+            if struct._is_scalar_type(object):
+                dtype_size = max(1, object.width // 8)
+                sub_align = max(dtype_size, sub_align)
+                max_size = max(max_size, dtype_size)
+            elif isinstance(object, struct._MemRangeMeta):
+                sub_align = max(object.elem_width // 8, sub_align)
+                max_size = max(max_size, object.size_in_bytes)
+            elif isinstance(object, struct):
+                sub_align = max(object.__alignof__(), sub_align)
+                max_size = max(max_size, object.__sizeof__())
+            else:
+                raise TypeError(
+                    f"Union element only support struct/array/base_dsl scalar, "
+                    f"but got {object}"
+                )
+            # Union alignment is the maximum alignment of all members
+            max_alignment = max(max_alignment, sub_align)
+
+        # Union size is the maximum size, aligned to the maximum alignment
+        self._align_of = max_alignment
+        self._size_of = struct.align_offset(max_size, max_alignment)
+
+    def __call__(self, base: Any) -> None:
```

Comment:
```suggestion
    @dsl_user_op
    def __call__(self, base: Any, *, loc=None, ip=None) -> None:
```

Reply:
`@dsl_user_op` does multiple things like generating location information. Let's add it here.